### PR TITLE
#[187364737] - An admin should be able to set roles and permissions

### PR DESCRIPTION
**Pull request by AngeloChristian1**
**What does this PR do**
Admin should be able to assign various users to various roles.

**Description of Task to be completed?**
- Created middleware to check for  user role
- Assigned permissions to various roles

 **How should this be manually tested?**
navigate to the `users/newRole/:userId` endpoint and add the id of a user you want to change
Make sure you have admin privileges 

https://www.loom.com/share/6d8f8b5be97248d1815c3761e739e64d?sid=fc378fd9-3c48-4bbe-9cc8-94c317b619e6

 **What are the relevant pivotal tracker/Trello stories?**
#187364737

### CODE REVIEW

File: src/controllers/UserController.ts [hunk 1]
```diff
@@ -42,3 +43,32 @@ export default class UserController {
...
+    }
+
+    const existingUser = await User.findOne({ where: { id: userId } })
+
+    if (!existingUser) {
+      return res.status(404).json({ message: 'User not found' })
+    }
+
+    existingUser.update({ userRole: newRole })
+    return res.status(201).json({
```
Comment (Yvan-David) [inline src/controllers/UserController.ts]: I think status code should not be 201  may be 200

File: test/middlewares.test.ts
Comment (Yvan-David) [inline test/middlewares.test.ts:1]: i think you should delete the file as its all commented

**Comment by Yvan-David** [comment]
please show how your feature works either by screenshots or a video

File: src/helpers/swagger.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import swaggerJsdoc, { Options } from 'swagger-jsdoc'
 
-const port = process.env.PORT || 8000 // Assuming you have PORT defined somewhere
+const port = process.env.PORT || 8080 // Assuming you have PORT defined somewhere
```
Comment (Furaha-Justine) [inline src/helpers/swagger.ts]: Why did you change the port number?

File: src/docs/loginDocs.ts
Comment (Furaha-Justine) [inline src/docs/loginDocs.ts:1]: I think to separate swagger docs files according to their functionality will reduce confusion. You can put this docs in its own file.

File: src/middlewares/index.ts
Comment (Furaha-Justine) [inline src/middlewares/index.ts:1]: Can you rename this file to something else.

**Review by Furaha-Justine — CHANGES_REQUESTED** [review]
Great work, but with these changes it will be much better.

File: src/controllers/UserController.ts [hunk 3]
```diff
@@ -34,3 +35,32 @@ export default class UserController {
...
+export const changeUserRole = async (req: Request, res: Response) => {
+  try {
+    const { userId } = req.params
+    const { newRole } = req.body
+
+    if (!userId || !newRole) {
+      return res.status(400).json({ message: 'Bad request' })
+    }
+
+    if (!['admin', 'buyer', 'seller'].includes(newRole)) {
```
Comment (munezerobagira) [inline src/controllers/UserController.ts]:
```suggestion
    if (![Role.ADMIN, Role.BUYER, Role.SELLER].includes(newRole)) {
```
We can use enum instead

File: src/routes/userRoute.ts [hunk 4]
```diff
@@ -1,14 +1,24 @@
...
 router.post(
   '/signup',
   validateRequest(userSchema, 'body'),
   UserController.signup,
 )
 
+router.post(
+  '/changeRole/:userId',
+  isAuthenticated,
+  checkPermission('admin'),
```
Comment (munezerobagira) [inline src/routes/userRoute.ts]: Use enum

**Review by munezerobagira — CHANGES_REQUESTED** [review]
Good job, but there are few things which we can improve

File: src/controllers/UserController.ts [hunk 5]
```diff
@@ -34,3 +35,32 @@ export default class UserController {
     })
   }
 }
+
+export const changeUserRole = async (req: Request, res: Response) => {
```
Comment (EricNiyo73) [inline src/controllers/UserController.ts]: put this function into the class , wherever that function can be  accessed using the class name as the previous functions from controllers looks like.
Remember that, static methods help  to organize code by associating a function with a class, rather than with individual objects like that.

**Review by EricNiyo73 — CHANGES_REQUESTED** [review]

**Review by EricNiyo73 — APPROVED** [review]

**Review by Furaha-Justine — APPROVED** [review]

**Review by munezerobagira — APPROVED** [review]
